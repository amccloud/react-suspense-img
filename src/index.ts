export { resource } from './resource';
export { Img } from './img';
