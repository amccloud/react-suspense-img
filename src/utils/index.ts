/**
 * Given a value, determine if it's a promise.
 *
 * @returns true if the value is a promise, false otherwise.
 *
 * @param promise the value to check if it's a promise
 */
export const isPromise = (promise: any): promise is Promise<any> =>
  !!promise && typeof promise.then === 'function';
