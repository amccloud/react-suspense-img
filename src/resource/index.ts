import invariant from 'invariant';

import { isPromise } from '../utils';

/**
 * The potential values that can be in the cache.
 *
 * - `Promise`: the image is loading.
 * - `Error`: an error occurred loading the image.
 * - `true`: the image was loaded successfully.
 */
type CacheValue = Promise<string> | Error | true;

interface ResourceCache {
  [src: string]: CacheValue;
}

/**
 * A simple resource to load images and suspend with React Concurrent mode.
 *
 * This is currently a pretty naive implementation that is inspired by a few
 * early demo apps and the Relay implementation. It has an infinite "cache"
 * size and relies on loading an image via JavaScript into the browser's cache.
 * Since the image itself is not stored in the internal resource's cache the
 * value is simply a boolean. There are likely many improvements that can be
 * made but without a lot of real world examples and use cases it's hard to
 * tell what needs to be more robust or where there may be edge cases.
 */
class Resource {
  private _cache: ResourceCache;

  constructor() {
    this._cache = {};
  }

  /**
   * Read an image by source from the cache. If it's cached, the source will
   * be returned. If it's not, a promise will be thrown. If it was attempted
   * but an error occurred an error will be thrown.
   *
   * @returns the source string if loaded, can also throw a promise and error.
   *
   * @param src the image source to read
   */
  public read = (src: string): string => {
    const cachedValue = this._cache[src];

    // 1. Check if the request to load the image is already in-flight or an
    // error has already occurred trying to load it.
    if (isPromise(cachedValue) || cachedValue instanceof Error) {
      throw cachedValue;
    }

    // 2. If it exists in the cache, the image has already been loaded.
    if (cachedValue === true) {
      return src;
    }

    // 3. It doesn't exist so it needs to be fetched. Ideally, this line is
    // never hit because that means the image was never preloaded.
    invariant(
      !cachedValue,
      `An image was loaded that wasn't preloaded.\n` +
        `Consider adding \`resource.preloadImage("${src}")\`` +
        ` before using with <Img src="${src}" />`
    );

    this.preloadImage(src);
    throw this._cache[src];
  };

  /**
   * Preload an image by source. Call this before trying to read an image.
   *
   * @param src the image source to load
   */
  public preloadImage = (src: string) => {
    const cachedValue = this._cache[src];

    // 1. Check if the request to load the image is already in-flight or an
    // error has already occurred trying to load it.
    if (isPromise(cachedValue) || cachedValue instanceof Error) {
      return cachedValue;
    }

    // 2. If it exists in the cache, the image has already been loaded.
    if (cachedValue === true) {
      return src;
    }

    const promise = new Promise<string>(async (resolve) => {
      const img = new Image();
      img.src = src;

      await image.decode();

      this._cache[src] = true;
      resolve(src);
    });

    this._cache[src] = promise;

    return promise;
  };

  /**
   * Clear every image from the resource cache.
   */
  public clear = () => {
    this._cache = {};
  };
}

// Initialize a single global resource for all images.
const resource = new Resource();

export { resource };
