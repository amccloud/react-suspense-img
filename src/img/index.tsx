import * as React from 'react';

import { resource } from '../resource';

type StandardImageProps = JSX.IntrinsicElements['img'];

interface Props extends StandardImageProps {
  src: string;
}

/**
 * Renders an image. If the image has not loaded yet it will suspend. If there
 * is an error while the image loads it will throw.
 *
 * @param props standard image attributes
 */
const Img = (props: Props) => {
  const { src, alt, ...rest } = props;

  const loadedSrc = resource.read(src);

  return <img {...rest} alt={alt} src={loadedSrc} />;
};

export { Img };
