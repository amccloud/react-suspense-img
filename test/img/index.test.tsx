import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';

import { Img, resource } from '../../src';

jest.mock('scheduler', () => require('scheduler/unstable_mock'));

interface Props {
  callback: (error: Error) => void;
}

interface State {
  hasError: boolean;
}

class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error: Error) {
    this.props.callback(error);
  }

  render() {
    if (this.state.hasError) {
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}

describe('<Img />', () => {
  let container: HTMLElement;
  let root: ReactDOM.Root;
  let onload: () => void;
  let onerror: () => void;

  beforeAll(() => {
    Object.defineProperty(Image.prototype, 'src', {
      set() {
        onload = () => this.onload();
        onerror = () => this.onerror();
      },
    });
  });

  beforeEach(() => {
    container = document.createElement('div');
    root = ReactDOM.createRoot(container);

    // Clear-out the image cache to start each test with a clean slate.
    resource.clear();
  });

  afterEach(() => {
    root.unmount();
  });

  it('suspends while loading the image', async () => {
    await act(async () =>
      root.render(
        <React.Suspense fallback={'Loading...'}>
          <Img src="https://placekitten.com/12/34" />
        </React.Suspense>
      )
    );

    expect(container.textContent).toBe('Loading...');

    await act(async () => onload());

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('throws an error if there was an issue loading the image', async () => {
    // Intentionally silencing verbose error messages from React, ReactDOM, etc.
    // since we're intentionally triggering an error.
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const callback = jest.fn();

    await act(async () =>
      root.render(
        <ErrorBoundary callback={callback}>
          <React.Suspense fallback={'Loading...'}>
            <Img src="https://placekitten.com/12/34" />
          </React.Suspense>
        </ErrorBoundary>
      )
    );

    await act(async () => onerror());

    expect(container.innerHTML).toMatchSnapshot();
    expect(callback).toHaveBeenCalledWith(
      Error(
        `An error occurred loading the image: "https://placekitten.com/12/34"`
      )
    );
  });

  it('passes through all standard image attributes', async () => {
    await act(async () =>
      root.render(
        <React.Suspense fallback={'Loading...'}>
          <Img
            src="https://placekitten.com/12/34"
            alt="kitten"
            width={64}
            title="A kitten"
          />
        </React.Suspense>
      )
    );
    await act(async () => onload());

    expect(container.innerHTML).toMatchSnapshot();
  });
});
