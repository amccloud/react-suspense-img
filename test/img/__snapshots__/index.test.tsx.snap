// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Img /> passes through all standard image attributes 1`] = `"<img width=\\"64\\" title=\\"A kitten\\" alt=\\"kitten\\" src=\\"https://placekitten.com/12/34\\">"`;

exports[`<Img /> suspends while loading the image 1`] = `"<img src=\\"https://placekitten.com/12/34\\">"`;

exports[`<Img /> throws an error if there was an issue loading the image 1`] = `"<h1>Something went wrong.</h1>"`;
